### PR TITLE
Support default arguments

### PR DESCRIPTION
**Pull request by kan-bayashi**
This PR support default args to define models.
This is useful to keep compatibility of the model when we add new options.
Also, this makes it easy to define model as follows:
```
[nav] In [1]: from espnet.nets.pytorch_backend.e2e_tts_tacotron2 import Tacotron2

[ins] In [2]: taco2 = Tacotron2(10, 10)
WARNING:root:attribute "embed_dim" does not exist. use default 512.
WARNING:root:attribute "elayers" does not exist. use default 1.
WARNING:root:attribute "eunits" does not exist. use default 512.
WARNING:root:attribute "econv_layers" does not exist. use default 3.
WARNING:root:attribute "econv_chans" does not exist. use default 512.
WARNING:root:attribute "econv_filts" does not exist. use default 5.
WARNING:root:attribute "atype" does not exist. use default location.
WARNING:root:attribute "adim" does not exist. use default 512.
WARNING:root:attribute "aconv_chans" does not exist. use default 32.
WARNING:root:attribute "aconv_filts" does not exist. use default 15.
WARNING:root:attribute "cumulate_att_w" does not exist. use default True.
WARNING:root:attribute "dlayers" does not exist. use default 2.
WARNING:root:attribute "dunits" does not exist. use default 1024.
WARNING:root:attribute "prenet_layers" does not exist. use default 2.
WARNING:root:attribute "prenet_units" does not exist. use default 256.
WARNING:root:attribute "postnet_layers" does not exist. use default 5.
WARNING:root:attribute "postnet_chans" does not exist. use default 512.
WARNING:root:attribute "postnet_filts" does not exist. use default 5.
WARNING:root:attribute "output_activation" does not exist. use default None.
WARNING:root:attribute "use_cbhg" does not exist. use default False.
WARNING:root:attribute "cbhg_conv_bank_layers" does not exist. use default 8.
WARNING:root:attribute "cbhg_conv_bank_chans" does not exist. use default 128.
WARNING:root:attribute "cbhg_conv_proj_filts" does not exist. use default 3.
WARNING:root:attribute "cbhg_conv_proj_chans" does not exist. use default 256.
WARNING:root:attribute "cbhg_highway_layers" does not exist. use default 4.
WARNING:root:attribute "cbhg_highway_units" does not exist. use default 128.
WARNING:root:attribute "cbhg_gru_units" does not exist. use default 256.
WARNING:root:attribute "use_batch_norm" does not exist. use default True.
WARNING:root:attribute "use_concate" does not exist. use default True.
WARNING:root:attribute "use_residual" does not exist. use default True.
WARNING:root:attribute "dropout_rate" does not exist. use default 0.5.
WARNING:root:attribute "zoneout_rate" does not exist. use default 0.1.
WARNING:root:attribute "reduction_factor" does not exist. use default 1.
WARNING:root:attribute "spk_embed_dim" does not exist. use default None.
WARNING:root:attribute "spc_dim" does not exist. use default None.
WARNING:root:attribute "use_masking" does not exist. use default False.
WARNING:root:attribute "bce_pos_weight" does not exist. use default 20.0.
WARNING:root:attribute "use_guided_attn_loss" does not exist. use default False.
WARNING:root:attribute "guided_attn_loss_sigma" does not exist. use default 0.4.
```

### CODE REVIEW

**Comment by kan-bayashi** [comment]
I'm not sure log is needed. 
If not needed, we can more simplify.

**Comment by kan-bayashi** [comment]
@sw005320 I think `AttributeDict` is not needed. Is there any special reason to use?

**Comment by sw005320** [comment]
> @sw005320 I think `AttributeDict` is not needed. Is there any special reason to use?

As far as I remember, I used this for the old-style model config file. I believe we don't need it anymore.

**Comment by kan-bayashi** [comment]
@sw005320 I found that `AttributeDict` return `None` when the specified attribute does not exist.
Is this your intension?

```
[ins] In [6]: args
Out[6]: <espnet.asr.asr_utils.AttributeDict at 0x7f0709f74cf8>

[ins] In [7]: args.aaaaaaaa

[ins] In [8]: import argparse

[ins] In [9]: args = argparse.Namespace(**dict(args.items()))

[ins] In [10]: args.aaaaaaaa
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-10-5a45b9b81631> in <module>
----> 1 args.aaaaaaaa

AttributeError: 'Namespace' object has no attribute 'aaaaaaaa’
```

**Comment by sw005320** [comment]
> @sw005320 I found that `AttributeDict` return `None` when the specified attribute does not exist.
> Is this your intension?
> 
> ```
> [ins] In [6]: args
> Out[6]: <espnet.asr.asr_utils.AttributeDict at 0x7f0709f74cf8>
> 
> [ins] In [7]: args.aaaaaaaa
> 
> [ins] In [8]: import argparse
> 
> [ins] In [9]: args = argparse.Namespace(**dict(args.items()))
> 
> [ins] In [10]: args.aaaaaaaa
> ---------------------------------------------------------------------------
> AttributeError                            Traceback (most recent call last)
> <ipython-input-10-5a45b9b81631> in <module>
> ----> 1 args.aaaaaaaa
> 
> AttributeError: 'Namespace' object has no attribute 'aaaaaaaa’
> ```
> 
> I think this is buggy, it is better to raise an error.

This is not my intention.
Yes, it's better to raise an error.

**Comment by kan-bayashi** [comment]
Done.

File: espnet/nets/pytorch_backend/e2e_tts_tacotron2.py [hunk 1]
```diff
@@ -283,94 +286,108 @@ class Tacotron2(TTSInterface, torch.nn.Module):
...
+        group.add_argument("--guided-attn-loss-sigma", default=0.4, type=float,
+                           help="Sigma in guided attention loss")
+        return parser
+
+    def __init__(self, idim, odim, args=None):
         # initialize base classes
         TTSInterface.__init__(self)
         torch.nn.Module.__init__(self)
 
+        # get default arguments and fill missing arguments
```
Comment (ShigekiKarita) [inline espnet/nets/pytorch_backend/e2e_tts_tacotron2.py]: I think we should split this to `espnet.net.args.py` or somewhere we can reuse this for the other models

File: espnet/nets/pytorch_backend/e2e_tts_transformer.py [hunk 2]
```diff
@@ -309,11 +314,20 @@ def attention_plot_class(self):
...
         return TTSPlot
 
-    def __init__(self, idim, odim, args):
+    def __init__(self, idim, odim, args=None):
         # initialize base classes
         TTSInterface.__init__(self)
         torch.nn.Module.__init__(self)
 
+        # get default arguments and fill missing arguments
+        args = {} if args is None else vars(args)
```
Comment (ShigekiKarita) [inline espnet/nets/pytorch_backend/e2e_tts_transformer.py]: ditto

**Comment by sw005320** [comment]
LGTM

**Pull request closed without merging**